codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/RustCrateInlineModuleComposingWriter.kt

```diff
@@ -0,0 +1,327 @@
+    return { symbol: Symbol, writable: Writable ->
+        this.withInlineModule(symbol.module()) {
+            writable()
+        }
+    }
+}
+
+/**
+ * Maintains the `RustWriter` that has been created for a `RustModule.LeafModule`.
+ */
+private data class InlineModuleWithWriter(val inlineModule : RustModule.LeafModule, val writer : RustWriter)
+
+/**
+ * For each RustCrate a separate mapping of inline-module to `RustWriter` is maintained.
+ */
+private val crateToInlineModule: ConcurrentHashMap<RustCrate, InnerModule> =
+    ConcurrentHashMap()
+
+class InnerModule(debugMode : Boolean) {
+    private val topLevelModuleWriters: MutableSet<RustWriter> = mutableSetOf()
+    private val inlineModuleWriters: ConcurrentHashMap<RustWriter, MutableList<InlineModuleWithWriter>> = ConcurrentHashMap()
+    private val docWriters: ConcurrentHashMap<RustModule.LeafModule, MutableList<DocWriter>> = ConcurrentHashMap()
+    private val writerCreator = RustWriter.factory(debugMode)
+    private val emptyLineCount: Int = writerCreator
+        .apply("lines-it-always-writes.rs", "crate")
+        .toString()
+        .split("\n")[0]
+        .length
+
+    fun withInlineModule(outerWriter: RustWriter, innerModule: RustModule.LeafModule, docWriter: DocWriter? = null, writable: Writable) {
+        if (docWriter != null) {
+            val moduleDocWriterList = docWriters.getOrPut(innerModule) { mutableListOf() }
+            moduleDocWriterList.add(docWriter)
+        }
+        writable(getWriter(outerWriter, innerModule))
+    }
+
+    /**
+     * Given a `RustCrate` and a `RustModule.LeafModule()`, it creates a writer to that module and calls the writable.
+     */
+    fun withInlineModuleHierarchyUsingCrate(rustCrate: RustCrate, inlineModule: RustModule.LeafModule, docWriter: DocWriter? = null, writable: Writable) {
+        val hierarchy = getHierarchy(inlineModule).toMutableList()
+        check(!hierarchy.first().isInline()) {
+            "When adding a `RustModule.LeafModule` to the crate, the topmost module in the hierarchy cannot be an inline module."
+        }
+        // The last in the hierarchy is the one we will return the writer for.
+        val bottomMost = hierarchy.removeLast()
+
+        // In case it is a top level module that has been passed (e.g. ModelsModule, OutputsModule) then
+        // register it with the topLevel writers and call the writable on it. Otherwise, go over the
+        // complete hierarchy, registering each of the inner modules and then call the `Writable`
+        // with the bottom most inline module that has been passed.
+        if (hierarchy.isNotEmpty()) {
+            val topMost = hierarchy.removeFirst()
+
+            // Create an intermediate writer for all inner modules in the hierarchy
+            rustCrate.withModule(topMost) {
+                var writer = this
+                hierarchy.forEach {
+                    writer = getWriter(writer, it)
+                }
```

Comment:
What does this ensure? Isn't at the end of this for loop `writer` equal to the last element in `hierarchy` (which is the penultimate module, since we called `removeLast` before).